### PR TITLE
Fixing datetime arithmetic overflow/underflow.

### PR DESCRIPTION
**Pull request by v-isyamauchi-gh**
## Description
Fix: Prevent DateTime overflow in search expression rewriters

FHIR search queries with extreme dates (e.g., _lastUpdated=gt9999-12-31, birthdate=ap9500-01-01) caused OverflowException in the SQL rewriting pipeline → 500 response. The overflow came from internal date arithmetic (AddTicks/AddDays) for SQL optimization, not invalid user input.

Changes:
- Added DateTimeSafeExtensions with SafeAddTicks/SafeAddDays that clamp to DateTime.Min/MaxValue instead of throwing
- Applied to LastUpdatedToResourceSurrogateIdRewriter, DateTimeBoundedRangeRewriter, ScalarTemporalEqualityRewriter, and SearchValueExpressionBuilderHelper (ap comparator)

## Related issues
Addresses [issue #181592].

[Bug 181592](https://microsofthealth.visualstudio.com/Health/_workitems/edit/181592): [Edge case] : 500 returned due to ArgumentOutOfRangeException for datetime

## Testing
Tested by adding some UTs.

## FHIR Team Checklist
- **Update the title** of the PR to be succinct and less than 65 characters
- **Add a milestone** to the PR for the sprint that it is merged (i.e. add S47)
- Tag the PR with the type of update: **Bug**, **Build**, **Dependencies**, **Enhancement**, **New-Feature** or **Documentation**
- Tag the PR with **Open source**, **Azure API for FHIR** (CosmosDB or common code) or **Azure Healthcare APIs** (SQL or common code) to specify where this change is intended to be released.
- Tag the PR with **Schema Version backward compatible** or **Schema Version backward incompatible** or **Schema Version unchanged** if this adds or updates Sql script which is/is not backward compatible with the code.
- When changing or adding behavior, if your code modifies the system design or changes design assumptions, please create and include an [ADR](https://github.com/microsoft/fhir-server/blob/main/docs/arch).
- [ ] CI is green before merge [Build Status](https://microsofthealthoss.visualstudio.com/FhirServer/_build/latest?definitionId=27&branchName=main) 
- Review [squash-merge requirements](https://github.com/microsoft/fhir-server/blob/main/SquashMergeRequirements.md)

### Semver Change ([docs](https://github.com/microsoft/fhir-server/blob/main/docs/Versioning.md))
Patch|Skip|Feature|Breaking (reason)

### CODE REVIEW

File: src/Microsoft.Health.Fhir.Core/Extensions/DateTimeSafeExtensions.cs [hunk 1]
```diff
@@ -0,0 +1,87 @@
...
+                return DateTime.MaxValue;
+            }
+
+            if (ticks < 0 && value.Ticks < DateTime.MinValue.Ticks - ticks)
+            {
+                return DateTime.MinValue;
+            }
+
+            return value.AddTicks(ticks);
+        }
```
[the remark this answers is not in the document]
Comment (v-isyamauchi-gh) [reply src/Microsoft.Health.Fhir.Core/Extensions/DateTimeSafeExtensions.cs]: Fixed

File: src/Microsoft.Health.Fhir.Core.UnitTests/Extensions/DateTimeSafeExtensionsTests.cs [hunk 2]
```diff
@@ -0,0 +1,152 @@
...
+
+        [Fact]
+        public void GivenDateTimeOffsetNearMinValue_WhenSafeAddDaysNegative_ThenClampsToMinValue()
+        {
+            var dto = DateTimeOffset.MinValue.AddDays(0.5);
+            DateTimeOffset result = dto.SafeAddDays(-1);
+            Assert.Equal(DateTimeOffset.MinValue, result);
+        }
+    }
+}
```
[the remark this answers is not in the document]
Comment (v-isyamauchi-gh) [reply src/Microsoft.Health.Fhir.Core.UnitTests/Extensions/DateTimeSafeExtensionsTests.cs:532]: Fixed

File: src/Microsoft.Health.Fhir.Core/Extensions/DateTimeSafeExtensions.cs [hunk 3]
```diff
@@ -0,0 +1,101 @@
...
+                return DateTimeOffset.MaxValue;
+            }
+
+            if (ticks < 0 && value.Ticks < DateTimeOffset.MinValue.Ticks - ticks)
+            {
+                return DateTimeOffset.MinValue;
+            }
+
+            return value.AddTicks(ticks);
+        }
```
[the remark this answers is not in the document]
Comment (v-isyamauchi-gh) [reply src/Microsoft.Health.Fhir.Core/Extensions/DateTimeSafeExtensions.cs]: Fixed

File: src/Microsoft.Health.Fhir.Core.UnitTests/Extensions/DateTimeSafeExtensionsTests.cs [hunk 4]
```diff
@@ -0,0 +1,230 @@
...
+        public void GivenDateTimeOffset_WhenSafeAddDaysClampsToMaxValue_ThenPreservesOffset()
+        {
+            var offset = TimeSpan.FromHours(5);
+            var dto = new DateTimeOffset(2024, 1, 1, 0, 0, 0, offset);
+            DateTimeOffset result = dto.SafeAddDays(int.MaxValue);
+            Assert.Equal(offset, result.Offset);
+            Assert.Equal(DateTimeOffset.MaxValue.Ticks, result.Ticks);
+        }
+    }
+}
```
[the remark this answers is not in the document]
Comment (v-isyamauchi-gh) [reply src/Microsoft.Health.Fhir.Core.UnitTests/Extensions/DateTimeSafeExtensionsTests.cs:532]: Fixed

File: src/Microsoft.Health.Fhir.Core.UnitTests/Extensions/DateTimeSafeExtensionsTests.cs [hunk 5]
```diff
@@ -0,0 +1,250 @@
...
+        {
+            var dt = new DateTime(2024, 1, 1, 0, 0, 0, DateTimeKind.Utc);
+            DateTime result = dt.SafeAddTicks(long.MinValue);
+            Assert.Equal(DateTime.MinValue, result);
+        }
+
+        [Fact]
+        public void GivenUtcDateTime_WhenSafeAddTicksClampsToMaxValue_ThenPreservesKind()
+        {
+            var dt = new DateTime(2024, 1, 1, 0, 0, 0, DateTimeKind.Utc);
```
[the remark this answers is not in the document]
Comment (v-isyamauchi-gh) [reply src/Microsoft.Health.Fhir.Core.UnitTests/Extensions/DateTimeSafeExtensionsTests.cs]: Fixed

File: src/Microsoft.Health.Fhir.R4.Web/Properties/launchSettings.json [hunk 6]
```diff
@@ -22,21 +22,21 @@
...
-        "FhirServer:Operations:IntegrationDataStore:StorageAccountConnection": "UseDevelopmentStorage=true",
-        "FhirServer:Operations:Import:Enabled": "true",
-        "ASPNETCORE_ENVIRONMENT": "development"
-      },
+        "environmentVariables": {
+            "SqlServer:Initialize": "true",
+            "SqlServer:AllowDatabaseCreation": "true",
+            "SqlServer:SchemaOptions:AutomaticUpdatesEnabled": "true",
+            "TestAuthEnvironment:FilePath": "..//..//testauthenvironment.json",
+            "SqlServer:ConnectionString": "server=V-IYAMAUCHI00\\MSSQLSERVER01;Initial Catalog=FHIR_R4;Integrated Security=true;TrustServerCertificate=true",
```
[the remark this answers is not in the document]
Comment (v-isyamauchi-gh) [reply src/Microsoft.Health.Fhir.R4.Web/Properties/launchSettings.json]: Fixed

File: src/Microsoft.Health.Fhir.Core/Features/Search/Expressions/Parsers/SearchValueExpressionBuilderHelper.cs [hunk 7]
```diff
@@ -102,8 +102,8 @@ void ISearchValueVisitor.Visit(DateTimeSearchValue dateTime)
 
                     var differenceTicks = (long)((Clock.UtcNow.Ticks - Math.Max(startTicks, endTicks)) * ApproximateMultiplier);
 
-                    var approximateStart = dateTime.Start.AddTicks(-differenceTicks);
-                    var approximateEnd = dateTime.End.AddTicks(differenceTicks);
+                    var approximateStart = dateTime.Start.SafeAddTicks(-differenceTicks);
+                    var approximateEnd = dateTime.End.SafeAddTicks(differenceTicks);
```
[the remark this answers is not in the document]
Comment (v-isyamauchi-gh) [reply src/Microsoft.Health.Fhir.Core/Features/Search/Expressions/Parsers/SearchValueExpressionBuilderHelper.cs:106]: Fixed

File: src/Microsoft.Health.Fhir.SqlServer/Features/Search/Expressions/Visitors/LastUpdatedToResourceSurrogateIdRewriter.cs [hunk 8]
```diff
@@ -56,7 +57,7 @@ public override Expression VisitBinary(BinaryExpression expression, object conte
                     return Expression.GreaterThanOrEqual(
                         SqlFieldName.ResourceSurrogateId,
                         null,
-                        new DateTimeOffset(truncated.AddTicks(TimeSpan.TicksPerMillisecond)).ToSurrogateId());
+                        new DateTimeOffset(truncated.SafeAddTicks(TimeSpan.TicksPerMillisecond)).ToSurrogateId());
```
[the remark this answers is not in the document]
Comment (v-isyamauchi-gh) [reply src/Microsoft.Health.Fhir.SqlServer/Features/Search/Expressions/Visitors/LastUpdatedToResourceSurrogateIdRewriter.cs:60]: It is a dupe of another one. Resolving this one.

File: src/Microsoft.Health.Fhir.Core/Features/Search/Expressions/Parsers/SearchValueExpressionBuilderHelper.cs (hunk 7 above)
[the remark this answers is not in the document]
Comment (v-isyamauchi-gh) [reply src/Microsoft.Health.Fhir.Core/Features/Search/Expressions/Parsers/SearchValueExpressionBuilderHelper.cs:106]: There are some UTs that are expecting without Math.Abs(differenceTicks). Need to confirm the meaning of without Abs.

[the remark this answers is not in the document]
Comment (v-isyamauchi-gh) [reply src/Microsoft.Health.Fhir.Core/Features/Search/Expressions/Parsers/SearchValueExpressionBuilderHelper.cs:106]: Resolve as by design.  Some cases of GivenADateWithApComparator_WhenBuilt_ThenCorrectExpressionShouldBeCreated test negative ticks cases explicitly, and the suggested changes will cause these cases to fail.

File: src/Microsoft.Health.Fhir.SqlServer/Features/Search/Expressions/Visitors/LastUpdatedToResourceSurrogateIdRewriter.cs (hunk 8 above)
[the remark this answers is not in the document]
Comment (v-isyamauchi-gh) [reply src/Microsoft.Health.Fhir.SqlServer/Features/Search/Expressions/Visitors/LastUpdatedToResourceSurrogateIdRewriter.cs:60]: Resolve as Won't Fix. 

We have an internal logic in [ToId](https://github.com/microsoft/fhir-server/blob/b88d5f7f66e6abdf3e09c4f2c930143c7684cbfb/src/Microsoft.Health.Fhir.Core/Extensions/IdHelper.cs#L27)() (called by ToSurrogateId()) to calculate a surrogate id (internal data) that limits the range of an input datetime offset value for its calculation. Any input outside of the limit will result in the 500 response. Addressing the issue appropriately will require some thought that is out of scope for this PR.

File: src/Microsoft.Health.Fhir.SqlServer/Features/Search/Expressions/Visitors/LastUpdatedToResourceSurrogateIdRewriter.cs [hunk 9]
```diff
@@ -45,43 +46,103 @@ public override Expression VisitBinary(BinaryExpression expression, object conte
...
+            }
+        }
+
+        private static BinaryExpression VisitBinaryConstrained(BinaryExpression expression)
+        {
+            // Constrain to MaxDateTime - Dates beyond MaxDateTime are outside the system's representable range.
+            // ResourceSurrogateId encodes the datetime in high bits and a uniquifier (0-7) in low 3 bits.
+            // Use the max surrogate ID for the constrained millisecond (uniquifier = 7) to correctly account for all rows in that bucket.
+            DateTime constrained = IdHelper.MaxDateTime.UtcDateTime;
+            long maxSurrogateId = new DateTimeOffset(constrained).ToSurrogateId() + 7;
```
Comment (LTA-Thinking) [inline src/Microsoft.Health.Fhir.SqlServer/Features/Search/Expressions/Visitors/LastUpdatedToResourceSurrogateIdRewriter.cs]: The max surrogate id is larger than this. When the date time is converted to a surrogate id it is first trimmed to milliseconds, converted to ticks, then bit shifted. Since one tick is 100ns, there are 10,000 ticks per millisecond. So there are 80,000 surrogate ids per millisecond, not 8. This needs to add 79,999.

**Comment by LTA-Thinking** [comment]
Could an E2E test be added that covers this scenario?

**Comment by LTA-Thinking** [comment]
I really wish we could get a logger to these conversions so we could know if the edge case has been triggered, but there aren't loggers available to any of the classes calling this. I wish there was a good way to fix this, but I don't see it without larger changes.

File: test/Microsoft.Health.Fhir.Shared.Tests.E2E/Rest/Search/DateSearchTests.cs [hunk 10]
```diff
@@ -299,6 +299,46 @@
...
+                }
+            }
+            catch (FhirClientException fce)
+            {
+                Assert.Fail($"A non-expected '{nameof(FhirClientException)}' was raised. Url: {Client.HttpClient.BaseAddress}. Activity Id: {fce.Response.GetRequestId()}. Error: {fce.Message}");
+            }
+            catch (Exception e)
+            {
+                Assert.Fail($"A non-expected '{e.GetType()}' was raised. Url: {Client.HttpClient.BaseAddress}. No Activity Id present. Error: {e.Message}");
+            }
```
[the remark this answers is not in the document]
Comment (v-isyamauchi-gh) [reply test/Microsoft.Health.Fhir.Shared.Tests.E2E/Rest/Search/DateSearchTests.cs:339]: Won't fix - follow the other cases.

File: src/Microsoft.Health.Fhir.SqlServer/Features/Search/Expressions/Visitors/LastUpdatedToResourceSurrogateIdRewriter.cs (hunk 9 above)
[the remark this answers is not in the document]
Comment (v-isyamauchi-gh) [reply src/Microsoft.Health.Fhir.SqlServer/Features/Search/Expressions/Visitors/LastUpdatedToResourceSurrogateIdRewriter.cs]: Fixed.

Comment (v-isyamauchi-gh) [reply src/Microsoft.Health.Fhir.SqlServer/Features/Search/Expressions/Visitors/LastUpdatedToResourceSurrogateIdRewriter.cs]: Fixed as we discussed offline

**Review by LTA-Thinking — DISMISSED** [review]

**Review by LTA-Thinking — DISMISSED** [review]

**Review by LTA-Thinking — APPROVED** [review]
